keras_nlp/models/gpt2/gpt2_causal_lm.py

```diff
@@ -356,28 +289,36 @@ def generate(
         """
         if self.preprocessor is None:
             raise ValueError(
-                "`self.preprocessor` is None, please make sure `preprocessor` "
-                "is set before calling `generate`."
+                "`self.preprocessor` is `None`, please make sure "
+                "`preprocessor` is set before calling `generate`."
             )
-        end_token_id = self.preprocessor.tokenizer.end_token_id
-
         sampler = keras_nlp.samplers.get(sampler)
-        if hasattr(self, "jit_compile"):
-            # `jit_compile` is a public property as of tf 2.12. hasattr is for
-            # backward compat.
-            sampler.jit_compile = self.jit_compile
+        # `jit_compile` is a property of keras.Model after tf 2.12.
+        # Use `getattr()` for backwards compatibility.
+        sampler.jit_compile = getattr(self, "jit_compile", True)
         sampler.run_eagerly = self.run_eagerly
-        x, _, _ = self.preprocessor(prompt)
-        if len(x["token_ids"].shape) == 1:
-            x["token_ids"] = x["token_ids"][tf.newaxis, :]
-            x["padding_mask"] = x["padding_mask"][tf.newaxis, :]
-        _, cache = self.build_initial_cache(x, max_length)
-        next_token_probability = self._get_token_probability
+
+        # Tokenize.
+        prompt = self.preprocessor.tokenizer(prompt)
+
+        # Create and seed the cache before generation.
+        token_ids = prompt
+        if prompt.shape.rank == 1:
+            token_ids = tf.RaggedTensor.from_tensor(prompt[tf.newaxis, :])
+        token_ids = token_ids.to_tensor(shape=(None, max_length))
+        padding_mask = tf.ones_like(token_ids, dtype=tf.bool)
```

Comment:
This is a dense tensor with all 1s. The functionality is unaffected because causal mask is used, but it might hurt our readability. Shall set the correct value of `padding_mask`, or add a comment here saying "the value of `padding_mask` does not matter because causal mask is applied"?

Reply:
I'll add the comment for now.

In future cleanup I would actually like the remove the mask updates we are doing in the sampler method, as they are totally unnecessary and just slow us down! Essentially the only masks we need to care about during generation are a causal mask and the mask of where we the original inputs were in our sequence tensor. Slowly scattering ones into our padding mask is just busywork.

I do struggle with how to make this clear to users though.